### PR TITLE
cowsay: fix commands for zsh

### PR DESCRIPTION
**Pull request by nicokosi**
Fix commands that fail with `zsh` due to missing escaping for the exclamation mark:

	$ cowsay -f dragon "Hello!"
	dquote>

	$ uname -a
	Darwin **** 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64

	$ echo $SHELL
	/bin/zsh

	$ cowsay -h
	cow{say,think} version 3.03, (c) 1999 Tony Monroe
	Usage: cowsay [-bdgpstwy] [-h] [-e eyes] [-f cowfile]
          [-l] [-n] [-T tongue] [-W wrapcolumn] [message]

- [x] The PR title conforms to the recommended [templates](/tldr-pages/tldr/blob/master/CONTRIBUTING.md#commit-message).
- [x] The page follows the [content guidelines](/tldr-pages/tldr/blob/master/CONTRIBUTING.md#guidelines).

### CODE REVIEW

**Comment by Waples** [comment]
Works fine for me :thinking: 

tested in `bash` and `fish` shells on different machines (Arch Linux, RaspberryPi Os).

```bash
unames:
Linux **** 5.11.13-arch1-1 #1 SMP PREEMPT Sat, 10 Apr 2021 20:47:14 +0000 x86_64 GNU/Linux
Linux **** 5.10.17-v7+ #1403 SMP Mon Feb 22 11:29:51 GMT 2021 armv7l GNU/Linux
```

```bash
$  cowsay -f dragon hello!                                                                                                                                                                                 10:27:12
 ________
< hello! >
 --------
      \                    / \  //\
       \    |\___/|      /   \//  \\
            /0  0  \__  /    //  | \ \
           /     /  \/_/    //   |  \  \
           @_^_@'/   \/_   //    |   \   \
           //_^_/     \/_ //     |    \    \
        ( //) |        \///      |     \     \
      ( / /) _|_ /   )  //       |      \     _\
    ( // /) '/,_ _ _/  ( ; -.    |    _ _\.-~        .-~~~^-.
  (( / / )) ,-{        _      `-.|.-~-.           .~         `.
 (( // / ))  '/\      /                 ~-. _ .-~      .-~^-.  \
 (( /// ))      `.   {            }                   /      \  \
  (( / ))     .----~-.\        \-'                 .~         \  `. \^-.
             ///.----..>        \             _ -~             `.  ^-`  ^-_
               ///-._ _ _ _ _ _ _}^ - - - - ~                     ~-- ,.-~
                                                                  /.-~
```

**Comment by nicokosi** [comment]
> Works fine for me 
> 
> tested in `bash` and `fish` shells on different machines (Arch Linux, RaspberryPi Os).
> 
> ```shell
> unames:
> Linux **** 5.11.13-arch1-1 #1 SMP PREEMPT Sat, 10 Apr 2021 20:47:14 +0000 x86_64 GNU/Linux
> Linux **** 5.10.17-v7+ #1403 SMP Mon Feb 22 11:29:51 GMT 2021 armv7l GNU/Linux
> ```
> 
> ```shell
> $  cowsay -f dragon hello!                                                                                                                                                                                 10:27:12
>  ________
> < hello! >
>  --------
>       \                    / \  //\
>        \    |\___/|      /   \//  \\
>             /0  0  \__  /    //  | \ \
>            /     /  \/_/    //   |  \  \
>            @_^_@'/   \/_   //    |   \   \
>            //_^_/     \/_ //     |    \    \
>         ( //) |        \///      |     \     \
>       ( / /) _|_ /   )  //       |      \     _\
>     ( // /) '/,_ _ _/  ( ; -.    |    _ _\.-~        .-~~~^-.
>   (( / / )) ,-{        _      `-.|.-~-.           .~         `.
>  (( // / ))  '/\      /                 ~-. _ .-~      .-~^-.  \
>  (( /// ))      `.   {            }                   /      \  \
>   (( / ))     .----~-.\        \-'                 .~         \  `. \^-.
>              ///.----..>        \             _ -~             `.  ^-`  ^-_
>                ///-._ _ _ _ _ _ _}^ - - - - ~                     ~-- ,.-~
>                                                                   /.-~
> ```



> Works fine for me 
> 
> tested in `bash` and `fish` shells on different machines (Arch Linux, RaspberryPi Os).
> 
> ```shell
> unames:
> Linux **** 5.11.13-arch1-1 #1 SMP PREEMPT Sat, 10 Apr 2021 20:47:14 +0000 x86_64 GNU/Linux
> Linux **** 5.10.17-v7+ #1403 SMP Mon Feb 22 11:29:51 GMT 2021 armv7l GNU/Linux
> ```
> 
> ```shell
> $  cowsay -f dragon hello!                                                                                                                                                                                 10:27:12
>  ________
> < hello! >
>  --------
>       \                    / \  //\
>        \    |\___/|      /   \//  \\
>             /0  0  \__  /    //  | \ \
>            /     /  \/_/    //   |  \  \
>            @_^_@'/   \/_   //    |   \   \
>            //_^_/     \/_ //     |    \    \
>         ( //) |        \///      |     \     \
>       ( / /) _|_ /   )  //       |      \     _\
>     ( // /) '/,_ _ _/  ( ; -.    |    _ _\.-~        .-~~~^-.
>   (( / / )) ,-{        _      `-.|.-~-.           .~         `.
>  (( // / ))  '/\      /                 ~-. _ .-~      .-~^-.  \
>  (( /// ))      `.   {            }                   /      \  \
>   (( / ))     .----~-.\        \-'                 .~         \  `. \^-.
>              ///.----..>        \             _ -~             `.  ^-`  ^-_
>                ///-._ _ _ _ _ _ _}^ - - - - ~                     ~-- ,.-~
>                                                                   /.-~
> ```

Thank you @Waples for testing. In fact, `cowsay -f dragon "hello!"` does not work with `zsh`.

File: pages.it/common/cowsay.md [hunk 1]
```diff
@@ -5,19 +5,19 @@
 
 - Stampa una mucca ASCII che dice "Hello world!":
 
-`cowsay "Hello world!"`
+`cowsay "Hello world\!"`
```
Comment (nicokosi) [inline pages.it/common/cowsay.md:8]:
```suggestion
`cowsay 'Hello world!'`
```
which works for `bash` and `zsh`

File: pages.it/common/cowsay.md [hunk 2]
```diff
@@ -5,19 +5,19 @@
...
 
 - Usa il testo da standard input per il fumetto:
 
-`echo "Ciao!" | cowsay`
+`echo "Ciao\!" | cowsay`
```
Comment (nicokosi) [inline pages.it/common/cowsay.md:12]:
```suggestion
`echo 'Ciao!' | cowsay`
```

File: pages.it/common/cowsay.md [hunk 3]
```diff
@@ -5,19 +5,19 @@
...
 
 - Elenca tutti i personaggi disponibili:
 
 `cowsay -l`
 
 - Stampa un drago ASCII che dice "Ciao!":
 
-`cowsay -f dragon "Ciao!"`
+`cowsay -f dragon "Ciao\!"`
```
Comment (nicokosi) [inline pages.it/common/cowsay.md:20]:
```suggestion
`cowsay -f dragon 'Ciao!'`
```

File: pages.ko/common/cowsay.md [hunk 4]
```diff
@@ -5,20 +5,20 @@
 
 - "Hello world!"라고 말하는 ASCII cow 출력:
 
-`cowsay "Hello world!"`
+`cowsay "Hello world\!"`
```
Comment (nicokosi) [inline pages.ko/common/cowsay.md:8]:
```suggestion
`cowsay 'Hello world!'`
```

File: pages.ko/common/cowsay.md [hunk 5]
```diff
@@ -5,20 +5,20 @@
...
 
 - 풍선에 표준 입력의 텍스트 사용:
 
-`echo "Hello!" | cowsay`
+`echo "Hello\!" | cowsay`
```
Comment (nicokosi) [inline pages.ko/common/cowsay.md:12]:
```suggestion
`echo 'Hello!' | cowsay`
```

File: pages.ko/common/cowsay.md [hunk 6]
```diff
@@ -5,20 +5,20 @@
...
 
 - 사용 가능한 모든 문자 나열:
 
 `cowsay -l`
 
 - "Hello!"라고 말하는 ASCII dragon 출력:
 
-`cowsay -f dragon "Hello!"`
+`cowsay -f dragon "Hello\!"`
```
Comment (nicokosi) [inline pages.ko/common/cowsay.md:20]:
```suggestion
`cowsay -f dragon 'Hello!'`
```

File: pages/common/cowsay.md [hunk 7]
```diff
@@ -3,22 +3,22 @@
 > Generate an ASCII character (by default a cow) saying or thinking something.
 > More information: <https://github.com/tnalpgge/rank-amateur-cowsay>.
 
-- Print an ASCII cow saying "Hello world!":
+- Print an ASCII cow saying "Hello world\!":
 
-`cowsay "Hello world!"`
+`cowsay "Hello world\!"`
```
Comment (nicokosi) [inline pages/common/cowsay.md:8]:
```suggestion
`cowsay 'Hello world!'`
```

File: pages/common/cowsay.md [hunk 8]
```diff
@@ -3,22 +3,22 @@
...
 
 - Use text from standard input for the balloon:
 
-`echo "Hello!" | cowsay`
+`echo "Hello\!" | cowsay`
```
Comment (nicokosi) [inline pages/common/cowsay.md:12]:
```suggestion
`echo 'Hello!' | cowsay`
```

File: pages/common/cowsay.md [hunk 9]
```diff
@@ -3,22 +3,22 @@
...
 
 - List all available characters:
 
 `cowsay -l`
 
 - Print an ASCII dragon saying "Hello!":
 
-`cowsay -f dragon "Hello!"`
+`cowsay -f dragon "Hello\!"`
```
Comment (nicokosi) [inline pages/common/cowsay.md:20]:
```suggestion
`cowsay -f dragon 'Hello!'`
```

**Comment by nicokosi** [comment]
Commit 06b6a5e fixes that commands for `zsh` (replacing double-quotes with single-quotes).

**Comment by marchersimon** [comment]
> Thank you @Waples for testing. In fact, `cowsay -f dragon "hello!"` does not work with zsh.

Can confirm. `!"` actually escapes any future exclamation marks for history expansion (`!!`, `!$`, ...), and so `zsh` thinks there's an unterminated quote.

**Comment by nicokosi** [comment]
(I did a `git push --force` in order to edit the commit message: I mentioned the `zsh` specificity) :innocent:

File: pages/common/cowsay.md [hunk 10]
```diff
@@ -3,22 +3,22 @@
 > Generate an ASCII character (by default a cow) saying or thinking something.
 > More information: <https://github.com/tnalpgge/rank-amateur-cowsay>.
 
-- Print an ASCII cow saying "Hello world!":
+- Print an ASCII cow saying "Hello world\!":
```
Comment (marchersimon) [inline pages/common/cowsay.md:6]: Why?

Comment (nicokosi) [reply pages/common/cowsay.md:6]: I forgot to change for single quotes. 😇

Comment (nicokosi) [inline pages/common/cowsay.md:6]:
```suggestion
- Print an ASCII cow saying `Hello world!`:
```

**Comment by navarroaxel** [comment]
Can we just remove the `!` from the text in the examples? :joy:

**Comment by marchersimon** [comment]
That would be the easy way 😄.

**Comment by nicokosi** [comment]
> Can we just remove the `!` from the text in the examples?

That's cheating. https://en.wikipedia.org/wiki/%22Hello,_World!%22_program :wink:

Why not? I don't care but the current fix (`'Hello world!'`) is the closest to the original examples.

**Comment by marchersimon** [comment]
How about we leave everything as is and add
> \> In the Z-Shell `!"` has to  be replaced with `\!"`

to the command description?

But then I'm wondering if that's even worth it. `!` has to be escaped almost always with the exception in bash, when it's at the end of a string. Adding that note for this might seem a bit too trivial.

**Comment by nicokosi** [comment]
> How about we leave everything as is and add
> 
> > > In the Z-Shell `!"` has to  be replaced with `\!"`
> 
> to the command description?
> 
> But then I'm wondering if that's even worth it. `!` has to be escaped almost always with the exception in bash, when it's at the end of a string. Adding that note for this might seem a bit too trivial.

IMHO, it's better to choose a simple example that works for all shells (I like `Hello world!` but `Hello world` is OK for me, too). :relaxed:

**Comment by marchersimon** [comment]
I think we should keep the examples as generic as possible and since `!"` isn't a problem of `cowsay` it would fit better in a `zsh` related page.

File: pages/common/cowsay.md [hunk 11]
```diff
@@ -3,22 +3,22 @@
 > Generate an ASCII character (by default a cow) saying or thinking something.
 > More information: <https://github.com/tnalpgge/rank-amateur-cowsay>.
 
-- Print an ASCII cow saying "Hello world!":
+- Print an ASCII cow saying `Hello world!`:
 
-`cowsay "Hello world!"`
+`cowsay 'Hello world!'`
 
 - Use text from standard input for the balloon:
```
Comment (sbrl) [inline pages/common/cowsay.md:10]: Perhaps this will read better?
```suggestion
- Read text from stdin for the balloon:
```

**Review by sbrl — COMMENTED** [review]
Hey there! I've got a minor suggestion below, but in terms of the exclamation merk issue I recommend that we remove the exclamation mark altogether. It can also pose an issue in Bash depending on the configuration of the shell if not quotes.

In tldr pages we aim for simplicity, and it sounds like including the exclamation mark would increase the complexity of the page.

**Review by owenvoke — APPROVED** [review]

**Comment by marchersimon** [comment]
@nicokosi I think we should move back to double quotes too.

**Comment by nicokosi** [comment]
> @nicokosi I think we should move back to double quotes too.

Done in e4125a20.

**Review by bl-ue — APPROVED** [review]

**Review by bl-ue — APPROVED** [review]
```
 _______ 
< LGTM! >
 ------- 
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


```

**Review by marchersimon — APPROVED** [review]

**Pull request merged**